File: C++/Dijkstras.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <algorithm>
#include <deque>
#include <queue>
#include <stack>
#include <set>
#include <string>
#include <string.h>
#include <array>
#include <unordered_map>
#include <unordered_set>
#include <iomanip>
using namespace std;
typedef long long ll;
typedef long double ld;
#define pb push_back

const int MaxN = 1e5+10;
vector<pair<ll, int>> g[MaxN];      // length, dest
ll dist[MaxN];
int parent[MaxN];
ll INF = 1e18;
int n, m;

    // O((N+M)logM)
void dijkstras(int start) {
    using T = pair<ll, int>;
    priority_queue<T, vector<T>, greater<T>> pq;
    bool visited[n];
    fill(dist, dist+n, INF);
    fill(visited, visited+n, false);
    dist[start] = 0;
    parent[start] = -1;
    
    pq.push({0, start});
    while (!pq.empty()) {
        T cur = pq.top();
        pq.pop();
        int node = cur.second;
        if (visited[node]) continue;
        visited[node] = true;
        
        for (auto i : g[node]) {
            if (visited[i.second]) continue;
            
            if (cur.first + i.first < dist[i.second]) {
                dist[i.second] = cur.first + i.first;
                parent[i.second] = node;
                pq.push({dist[i.second], i.second});
            }
        }
    }
}

    // O(N^2)
void dijkstras2(int start) {
    bool visited[n];
    fill(dist, dist+n, INF);
    fill(visited, visited+n, false);
    dist[start] = 0;
    parent[start] = -1;
    
    while (true) {
        // find the smallest one
        int smallest=-1;
        ll minval=INF;
        for (int j = 0; j < n; j++) {
            if (!visited[j] && dist[j] < minval) {
                minval = dist[j];
                smallest = j;
            }
        }
        if (smallest == -1) break;
        
        for (auto a : g[smallest]) {
            if (!visited[a.second] && dist[a.second] > minval + a.first) {
                dist[a.second] = minval + a.first;
                parent[a.second] = smallest;
            }
        }
        visited[smallest] = true;
    }
}

vector<int> Backtrack(int dest) {
    vector<int> path;
    if (dist[dest] == INF) return path;
    while (dest != -1) {
        path.push_back(dest);
        dest = parent[dest];
    }
    return path;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> n >> m;

    for (int i = 0; i < m; i++) {
        int a, b; ll weight;
        cin >> a >> b >> weight;
        a--; b--;

        g[a].pb({weight, b});
        g[b].pb({weight, a});
    }
}

File: Java/Dijkstras.java
import java.util.*;
import java.io.*;

public class Dijkstras {
	
	static ArrayList<Edge>[] g;
	static long[] dist;
	static int[] parent;
	static int n, m;
	static final long INF = (long)(1e18);

	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		//BufferedReader in = new BufferedReader(new FileReader("Dijkstras"));

		StringTokenizer st = new StringTokenizer(in.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		dist = new long[n];
		parent = new int[n];
		g = new ArrayList[n];
		
		for (int i = 0; i < n; i++) {
			g[i] = new ArrayList<>();
		}
		
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(in.readLine());
			int a = Integer.parseInt(st.nextToken()) - 1; 	
			int b = Integer.parseInt(st.nextToken()) - 1; 	
			long c = Long.parseLong(st.nextToken()); 
			g[a].add(new Edge(b, c));
			g[b].add(new Edge(a, c));
		}
		
	}
	
		// O((N+M)logN)
	public static void dijkstras(int start) {
		PriorityQueue<Edge> pq = new PriorityQueue<>();
		boolean[] visited = new boolean[n];
		Arrays.fill(dist, INF);
		dist[start] = 0;
		parent[start] = -1;
		
		pq.add(new Edge(start, 0));
		while (!pq.isEmpty()) {
			Edge cur = pq.poll();
			int node = cur.dest;
			if (visited[node]) continue;
			visited[node] = true;
			
			for (Edge i : g[node]) {
				if (visited[i.dest]) continue;
				
				if (cur.length + i.length < dist[i.dest]) {
					dist[i.dest] = cur.length + i.length;
					parent[i.dest] = node;
					pq.add(new Edge(i.dest, dist[i.dest]));
				}
			}
		}
	}
	
		// O(N^2)
	public static void dijkstras2(int start) {
		boolean[] visited = new boolean[n];
		Arrays.fill(dist, INF);
		dist[start] = 0;
		parent[start] = -1;
		
		while (true) {
			// find the smallest one
			int smallest = -1;
			long minval = INF;
			for (int j=0; j<n; j++) {
				if (!visited[j] && dist[j] < minval) {
					minval = dist[j];
					smallest = j;
				}
			}
			if (smallest == -1) break;
			
			for (Edge a : g[smallest]) {
				if (!visited[a.dest] && dist[a.dest] > minval + a.length) {
					dist[a.dest] = minval + a.length;
					parent[a.dest] = smallest;
				}
			}
			visited[smallest] = true;
		}
	}
	
	public static ArrayList<Integer> Backtrack(int dest) {
		ArrayList<Integer> path = new ArrayList<>();
		if (dist[dest] == INF) {
			return path;
		}
		while (dest != -1) {
			path.add(dest);
			dest = parent[dest];
		}
		return path;
	}
	
	static class Edge implements Comparable<Edge> {
		int dest;
		long length;
		
		Edge(int dest , long len) {
			this.dest = dest;
			length = len;
		}
		
		public int compareTo(Edge o) {
			return Long.compare(length, o.length);
		}
	}
}
